### PR TITLE
Signup: Properly suggest non-taken usernames when signing up.

### PR DESCRIPTION
**Pull request by bisko**
This PR tries to improve the problem described in #3388 

Testing instructions can be found below.

While trying to fix the issue, I stumbled on a problem that I am not sure how to solve: 

In order to do the validity check, we need to ping the API to see if the user is available and to eventually get suggestions for free usernames. The call to the API can be a bit slow sometimes - 500ms to up to a few seconds, depending on network condition, etc. 

I tried to work out a scenario where this wouldn't be a problem for the user and got four ideas:

* Block the loading of the form and wait for the server to reply with a suggestion or with an OK message.
 * This would be very slow, compared to now, where it is instantaneous (there are no calls to the API made)
* Do the call to the API when the form is already loaded and a suggestion is filled in the field.
 * The shortcoming here is that the user will see one name that will possibly change to something else after a bit of time (to make the API call) or will see an error if none of the suggestions were valid.
* Do the call when the form is loaded and have a little loader on the field. The field will stay disabled/loading until the call is completed and it will get filled after that.
 * I am not sure if we have such a control (field with an inline loader) 
 * Disabling the field for the user would not be very optimal
* Get the suggestions when choosing the domain name. As the suggested username is based on the domain it would be logical to start the process to fetch the suggested names there.
 * This will allow for the call to finish before rendering the form.
 * Creates a dependency on another step. If the step is not done (for example if we decide to reorder the steps and choose a domain after that, or if we remove the domain choice entirely), the call will never be made. It is possible to verify if we already have the data and if not, to fetch it, but this is the same as solving the options above.

Currently I have implemented the second option above, to prove my code works. 

Testing instructions:

* Apply PR code
* Apply D1691-code to your sandbox
* Start signup
* Fill up domain `kwight` (as in the issue)
* Choose a domain, e.g. `kwight.me`
* Choose no GAPPs integration
* Wait a moment to get a username suggestion. 

cc @michaeldcain @coreh @meremagee @ehg @shaunandrews

### CODE REVIEW

**Comment by ehg** [comment]
From a cursory glance:
- The code is in the wrong place. We shouldn't be editing the signup-form component to do such dynamic operations, it should be handled in the `user` step of signup, and by the passing of props from the stores.
- The endpoint call should be in a signup action: https://github.com/Automattic/wp-calypso/blob/fix/3388-signup-suggest-free-username/client/lib/signup/step-actions.js

https://github.com/Automattic/wp-calypso/blob/master/client/signup/README.md :)

**Comment by bisko** [comment]
Moved the code to the `step-actions.js`.

The action is now fired on the completion of the `domains` step. This is done as early as possible, because the call to the API takes a bit of time and the `plans` step, in between the `domains` and the `user` steps, serves as an invisible delay to give the request a chance to finish. 

The resulting suggested username is stored in `localStorage` via `store`, in order to be used when the form gets initialised. If there was no suggested username (failed request or the request didn't manage to finish), the `username` field will remain empty.

The old way of suggesting usernames was removed and now the suggested username is based on only the site address (either `.wordpress.com` address or a custom domain), chosen in the domains step. 

cc @michaeldcain

**Comment by apeatling** [comment]
@bisko This one is gathering some dust, could you reping some folks on your team for review and to push it forward?

**Comment by michaeldcain** [comment]
@apeatling: this is on me, not @bisko. I'll get it reviewed today.

**Comment by michaeldcain** [comment]
We're researching a better way than using `store` to pass the suggested username between steps.

File: client/signup/steps/domains/index.jsx [hunk 1]
```diff
@@ -18,7 +18,8 @@ var StepWrapper = require( 'signup/step-wrapper' ),
 	Notice = require( 'components/notice' ),
 	abtest = require( 'lib/abtest' ).abtest,
 	signupUtils = require( 'signup/utils' ),
-	i18n = require( 'lib/mixins/i18n' );
+	i18n = require( 'lib/mixins/i18n' ),
```
Comment (Tug) [inline client/signup/steps/domains/index.jsx]: Hi there, i18n usage has changed recently. please use:
```
import i18n from 'i18n-calypso';
```
instead.

Comment (bisko) [inline client/signup/steps/domains/index.jsx]: Thanks for the tip! The library was already included above, so I removed that reference.

**Comment by bisko** [comment]
I tried rebasing the PR on #6124 and broke some things in the process, which resulted the PR to close itself. Will open a new PR when the code has been properly migrated to use the Redux store.

**Pull request closed without merging**